### DISCUSSION TITLE
new format: Confluence

### DISCUSSION THREAD

**Discussion by andrewcstewart** [quarto-dev/quarto-cli#2003] [Feature Requests]
Looking under the list of formats under https://quarto.org/docs/output-formats/all-formats.html, one format that would be nice to add is Confluence's.  I think there are probably a few different options there, but one would be https://confluence.atlassian.com/doc/confluence-storage-format-790796544.html 

Just starting this conversation to see if there is similar interest out there.

**Comment by jjallaire** [comment]
Yes, we are absolutely planning on building a Confluence format! 

cc @allenmanning

**Comment by ghost** [comment]
Thanks for reaching out.  I hope to start some work on this soon.  I've created [this issue](https://github.com/quarto-dev/quarto-cli/issues/2038) to track the initial work.

**Comment by zachcp** [comment]
@allenmanning  : I tried the 1.3 preview with Confluence support. It is truly fantastic, bravo sir. 

Question:  do you think there will ever be support for OJS within Confluence or that this will remain a strictly non-JS publishing route? For a number of analytical workflows I've moved to "aggregate the heavy stuff in a process" then do display in QMD/OJS. So if no OJS is available  it just means I'd need to put R/python back in the loop in order to generate static assets. No prob but curious as to the path forward.  Thanks again for your excellent work.

**Reply by ghost** [reply]
Thanks for your positive feedback.  We really hope you find it useful and I very much appreciate you sharing thoughts on around potential future features.

We can explore many different additions as we move forward.  One very important factor is the [limitations of CSF and what they will allow us to do](https://confluence.atlassian.com/doc/confluence-storage-format-790796544.html).

**Reply by msummersgill** [reply]
Excited to see this capability being developed! Is there any line of sight to potentially supporting HTML Widgets in the rendering/publishing process? Would love to be able to include content like plotly charts or interactive datatables, but from what I can tell, it appears that adding raw HTML to confluence pages currently requires paid add-ins.

**Reply by ghost** [reply]
As part of 1.3, there will be the ability to add Confluence Raw Blocks.  This will let you "pass-through" straight CSF to Confluence.

This should give us the ability to use the standard Confluence elements like Task Lists and Page Trees or just Raw HTML.

This gives us more flexibility.  At the end of the day, we are limited by what Confluence supports in their run-time.  

````
---
title: "Simple Doc - Confluence Raw Block"
format: confluence-html
---

# Tasks

```{=confluence}
<ac:task-list>
    <ac:task>
        <ac:task-id>1</ac:task-id>
        <ac:task-status>complete</ac:task-status>
        <ac:task-body><span class="placeholder-inline-tasks">Wake Up</span></ac:task-body>
    </ac:task>
    <ac:task>
        <ac:task-id>2</ac:task-id>
        <ac:task-status>incomplete</ac:task-status>
        <ac:task-body><span class="placeholder-inline-tasks">Brush Teeth</span></ac:task-body>
    </ac:task>
    <ac:task>
        <ac:task-id>3</ac:task-id>
        <ac:task-status>incomplete</ac:task-status>
        <ac:task-body><span class="placeholder-inline-tasks">Eat Breakfast</span></ac:task-body>
    </ac:task>
</ac:task-list>
```

# Emoji
```{=confluence}
<ac:emoticon ac:name="blue-star" ac:emoji-shortname=":sunrise_over_mountains:" ac:emoji-id="1f304" ac:emoji-fallback="🌄" />
```

# Decisions
```{=confluence}
<ac:adf-extension>
    <ac:adf-node type="decision-list">
        <ac:adf-attribute key="local-id">ad7b0a98-1e3e-4863-8f23-ff44dbd7dfa1</ac:adf-attribute>
        <ac:adf-node type="decision-item">
            <ac:adf-attribute key="local-id">c25ea574-62d6-4770-bf11-5a27f3309d56</ac:adf-attribute>
            <ac:adf-attribute key="state">DECIDED</ac:adf-attribute>
            <ac:adf-content>Use Quarto</ac:adf-content>
        </ac:adf-node>
    </ac:adf-node>
    <ac:adf-fallback>
        <ul class="decision-list">
            <li>Use Quarto</li>
        </ul>
    </ac:adf-fallback>
</ac:adf-extension>
```

# Page Tree

```{=confluence}
 <ac:structured-macro ac:name="pagetree" ac:schema-version="1" ac:local-id="19b56ab5-7126-4b5c-bb75-03763218ffc4" ac:macro-id="2889812da02196ef0d0256f347599000">
        <ac:parameter ac:name="root">
            <ac:link>
                <ri:page ri:content-title="@self" />
            </ac:link>
        </ac:parameter>
        <ac:parameter ac:name="startDepth">2</ac:parameter>
    </ac:structured-macro>
```


````

**Reply by fkgruber** [reply]
thanks
they all work except the page tree. Is this suppose to show a table of content?

**Reply by fkgruber** [reply]
Oh I see how it works how. It creates a table of content of subpages not of heading of the same page.
How would I modify this to create a table of content of headings within the same confluence page?

**Comment by ghost** [comment]
Here are the 1.3 Pre-release docs for Confluence: https://quarto.org/docs/prerelease/1.3/confluence.html

**Comment by andrewcstewart** [comment]
@allenmanning Is this intended to be primarily for Confluence Cloud or also self-hosted Confluence Server?

I ask because I notice there seem to be quite a few hard assumptions about things like URL structure and authentication method:

- not all sites use `/wiki` as the base path https://github.com/quarto-dev/quarto-cli/blob/6ec1848d8c2cbcff88ef511e496aa86f673322be/src/publish/confluence/confluence-helper.ts#L165
- Auth is requiring email address + access token, but are those being base64 encoded somewhere in-transit?  Because `quarto publish confluence` gives me an auth error on my confluence site, but I can hit the API directly when I generated a base64 encoded string from my email and access token.
- https://github.com/quarto-dev/quarto-cli/blob/6ec1848d8c2cbcff88ef511e496aa86f673322be/src/publish/confluence/confluence.ts#L100 also seems to assume that CONFLUENCE_DOMAIN, CONFLUENCE_USER_EMAIL and CONFLUENCE_AUTH_TOKEN are all supplied

**Reply by ghost** [reply]
We made this decision early on in the development to only focus on Confluence Cloud as Atlassian is EoLing Confluence Server in favor of Cloud.

More information on this can be found here: https://www.atlassian.com/migration/assess/journey-to-cloud.

Although Server is being EoLed, it does seem that there will be some long-standing installations of Confluence Data Center which will continue on without a Cloud migration.

[In this issue we are considering](https://github.com/quarto-dev/quarto-cli/issues/5072) adding support for Data Center. As we build up a potential case for this, we will link in issues as they are posted.

**Reply by andrewcstewart** [reply]
Thanks @allenmanning  - totally makes sense to not support Confluence Server (I would certainly make the same decision - it's a very messy product).

**Comment by maptv** [comment]
Can the Quarto CLI pull documents from Confluence, converting the documents to local qmd files in a Quarto Confluence project? I could imagine a collaboration workflow in which someone edits in Confluence and I run a command to pull in their changes, make my own edits on top of theirs and then run quarto publish confluence

**Reply by ghost** [reply]
@maptv Thanks for suggesting this idea.  I too can see how this could be useful.  The support is not currently there and I've captured it in this issue for consideration.  https://github.com/quarto-dev/quarto-cli/issues/5676

**Comment by zachcp** [comment]
Hi Team Confluence, 

Thanks again for this great work.   Is the currently a way to create page-level Table of Contents? Perhaps specified at the Project level? 

zach cp

**Reply by ghost** [reply]
Thanks for posting Zach.

Project index pages currently use a [Confluence Raw Block](https://quarto.org/docs/publishing/confluence.html#raw-confluence-blocks) to render a [Page Tree Macro](https://support.atlassian.com/confluence-cloud/docs/insert-the-page-tree-macro/).  This approach can be used in any QMD today and it might address what you are looking for.

Do you like the idea of wrapping this into a [confluence-specific shortcode](https://github.com/quarto-dev/quarto-cli/issues/5699)?

**Comment by DerIltis** [comment]
Hi everyone,

thanks for this great new feature. We would like to use it to replace conflR. However, we have some technical issues at the moment, and I would ask for your support here.

When I want to publish a small qmd file to our confluence via the console, it stops with the following fatal alert:

$ quarto publish confluence LMT_UserGuide.qmd
A new release of Deno is available: 1.28.2 → 1.34.1 Run `deno upgrade` to install it.
? Confluence Domain: » https://confluence.xfab.ads/
Sending fatal alert BadCertificate
[>] Verifying...
ERROR: https://confluence.xfab.ads/ doesn't exist

Error: https://confluence.xfab.ads/ doesn't exist
    at verifyLocationExists (file:///C:/Program%20Files/Quarto/bin/quarto.js:119821:15)
    at async withSpinner (file:///C:/Program%20Files/Quarto/bin/quarto.js:56613:9)
    at async verifyWithSpinner (file:///C:/Program%20Files/Quarto/bin/quarto.js:119798:12)
    at async verifyLocation (file:///C:/Program%20Files/Quarto/bin/quarto.js:119811:12)
    at async Object.promptAndAuthorizeToken [as authorizeToken] (file:///C:/Program%20Files/Quarto/bin/quarto.js:119886:5)
    at async resolveAccount (file:///C:/Program%20Files/Quarto/bin/quarto.js:121099:21)
    at async doPublish (file:///C:/Program%20Files/Quarto/bin/quarto.js:121397:65)
    at async publishAction (file:///C:/Program%20Files/Quarto/bin/quarto.js:121435:13)
    at async Command.fn (file:///C:/Program%20Files/Quarto/bin/quarto.js:121387:9)
    at async Command.execute (file:///C:/Program%20Files/Quarto/bin/quarto.js:8437:13)

We had this kind of issues in the past with conflR as well, but there we could disable ssl checks via:

library(httr)
set_config( config( ssl_verifypeer = 0L ) )
confl_create_post_from_Rmd("LMT_UserGuide.Rmd")

Of course I know that we should somehow resolve the root cause for this issue (What exactly could I do?).
However, I was wondering if there is a possibility to (temporarily) disable the ssl check for https connections in quarto?

**Reply by DerIltis** [reply]
Ah, I just realized you do not support Confluence Server (this is what we have here...), should / could this be the problem?

**Reply by ghost** [reply]
@DerIltis Thanks for posting.  We are tracking Support for Confluence Data Center [in this epic](https://github.com/quarto-dev/quarto-cli/issues/5072).  I will link in this comment.

**Comment by jjallaire** [comment]
In terms of the BadCertificate error, we just published an updated pre-release of v1.4 that should resolve this (https://quarto.org/docs/download/prerelease.html). We'll be back porting that fix to v1.3 (cc @dragonstyle). 

But once you get past that, yes, we don't yet support Confluence Server (although we are currently investigating what it would take to do that).

**Comment by mcanouil** [comment]
Closing as resolved and in favour of https://github.com/quarto-dev/quarto-cli/issues/2038
